api/src/opentrons/protocol_api/core/engine/module_core.py

```diff
@@ -156,19 +165,33 @@ def engage_to_labware(
             preserve_half_mm: For labware whose definitions
                 erroneously use half-mm for their defined default engage height,
                 use the value directly instead of converting it to real millimeters.
-
-        Raises:
-            Exception: Labware is not loaded or has no default engage height.
         """
-        raise NotImplementedError("engage_to_labware not implemented")
+        model = self._engine_client.state.modules.get_model(module_id=self.module_id)
+
+        default_height = (
+            self._engine_client.state.geometry.get_default_magnet_engage_height(
+                module_id=self.module_id,
+                preserve_half_mm=preserve_half_mm,
+            )
+        )
+
+        calculated_height = self._engine_client.state.modules.calculate_magnet_height(
+            module_model=model,
+            labware_default_height=default_height,
+            offset_from_labware_default=offset,
+        )
```

Comment:
These three selector calls can be combined into one call if we modify. `get_default_magnet_engage_height` a little We can do this because:

- We're not worrying about half-mm in the engine core, the engine itself will deal with that in the command implementation
    - This means we don't need `module_model`
- With the arguments we're using, all that `calculate_magnet_height` is doing is added `offset` to `default_height`
    - This _also_ means we don't need `module_model`
- Any information from the `ModulesView` or `LabwareView` that we need, the `GeometryView` can compose itself

Really, all we're doing is:

1. Get the engage height in `mm` from the labware definition
2. Add `offset` to that engage height
3. Pass that engage height to the `engage` command

The selectors we call should communicate that intent

```suggestion
        calculated_height = (
            self._engine_client.state.geometry.get_default_magnet_engage_height(
                module_id=self.module_id,
                offset_from_labware_default=offset,
            )
        )
```